mastodon/Mastodon.py

```diff
@@ -1017,19 +1035,39 @@ def __api_request(self, method, endpoint, params = {}, files = {}, do_ratelimiti
 
         return response
 
-    def __stream(self, endpoint, listener, params = {}):
+
+    def __stream(self, endpoint, listener, params = {}, async=False):
         """
         Internal streaming API helper.
+
+        Returns the requests.Response instance corresponding to the open websocket
+        connection.
         """
 
         headers = {}
         if self.access_token != None:
             headers = {'Authorization': 'Bearer ' + self.access_token}
-
         url = self.api_base_url + endpoint
-        with closing(requests.get(url, headers = headers, data = params, stream = True)) as r:
-            listener.handle_stream(r.iter_lines())
 
+        connection = requests.get(url, headers = headers, data = params, stream = True)
+
+        def __stream_threadproc():
+            with closing(connection) as r:
+                try:
+                    listener.handle_stream(r.iter_lines())
+                except AttributeError as e:
+                    # TODO If the user closes the connection early, requests gets
+                    # confused and throws an AttributeError
+                    pass
+            return 0
+
+        if async:
+            t = threading.Thread(args=(), target=__stream_threadproc)
+            t.start()
+            return connection
```

Comment:
I have considered just killing the thread but a) it felt possibly unsafe (if requests has any internal state that might be abandoned by a thread kill) and b) it can only be done from within the thread, so I would have to wait for the server to send a heartbeat so that I actually end up in my own code.

I considered creating a small wrapper class that passes along the close() but it felt like too much additional code for a small benefit. 

The thing is that I can't find any documentation on what the appropriate way to close a long polling/chunked response in Requests actually is. All of the search results on it are about the addition of the feature, and no one seems to have usage code that actually ever exits from handling a chunked response before the server is ready.

I investigated trying to indicate from the clientside that the connection is ready to be terminated (so that the server would initiate the connection drop) but I was unable to figure out how to send the zero-length termination message from requests' API. Hence why I was eager to find someone who was more familiar with the library.